### PR TITLE
[Insights] Introduce announced breaking changes, implement ShouldProcess, rename parameters to singluar names and add aliases to keep it non-breaking

### PR DESCRIPTION
**Pull request by gucalder**
## Description

* Deprecation of two cmdlets as announced in April 2017: Get-AzureRmUsage (API not being used), Add-AzureRmLogAlertRule (functionality moved to Azure ActivityLogAlerts API introduced in Ignite)
* Change the output type of the following cmdlets as announced in April 2017: Get-AzureRmAlertRule (flattening output,) Remove-AzureRmAlertRule (from list to single object,) Get-AzurRmAutoscaleSetting (minor change,) Remove-AzureRmLogProfile (from Boolean to object,) and Get-AzureRmLog, Get-AzureRmAlertHistory, and Get-AzureRmAutoscaleHistory (minor changes)
* Add the ShouldProcess functionality to these cmdlets Add-AzureRmMetricAlertRule, Add-AzureRmWebtestAlertRule, Remove-AzureRmAlertRule, Add-AzureRmAutoscaleSetting, Remove-AzureRmAutoscaleSetting, Add-AzureRmLogProfile, and Remove-AzureRmLogProfile. Remove the cmdlets from the exceptions list. Keep exceptions for the Force parameter.
* Add aliases (i.e. singular nouns) to those parameters that are currently using plural nouns. Also add deprecation warnings for the plural names.

This replaces PR https://github.com/Azure/azure-powershell/pull/4873/ issued from AuxMon organization.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by gucalder** [comment]
I think there are two empty lines at the end of the file Exceptions\BreakingChangeIssues.csv and they are breaking the StaticAnalysis.

**Comment by gucalder** [comment]
I do not know what the issue was this time. It failed very quickly.

File: src/ResourceManager/Insights/Commands.Insights.Test/Utilities.cs [hunk 1]
```diff
@@ -172,12 +172,12 @@ public static IEnumerable<MetricDefinition> InitializeMetricDefinitionResponse()
...
         }
 
-        public static void VerifyDetailedOutput(LogsCmdletBase cmdlet, ref string selected)
+        public static void VerifyDetailedOutput(LogsCmdletBase cmdlet)
         {
             // Calling with detailed output
             cmdlet.DetailedOutput = true;
             cmdlet.ExecuteCmdlet();
-            Assert.Null(selected); // Incorrect nameOrTargetUri clause with detailed output on
+            // Assert.Null(selected); // Incorrect nameOrTargetUri clause with detailed output on
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights.Test/Utilities.cs]: @gucalder please remove this line

File: src/ResourceManager/Insights/Commands.Insights/ActivityLogs/GetAzureRmLogCommand.cs [hunk 2]
```diff
@@ -104,7 +104,8 @@ public class GetAzureRmLogCommand : LogsCmdletBase
         /// <returns>The query filter with the conditions for particular parameters added</returns>
         protected override string ProcessParticularParameters(string currentQueryFilter)
         {
-            this.SetMaxEventsIfPresent(currentQueryFilter, this.MaxEvents);
+            WriteWarning("Parameter name change: The parameter plural names for the parameters will be deprecated in May 2018 in favor of the singular versions of the same names.");
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/ActivityLogs/GetAzureRmLogCommand.cs]: @gucalder please update the warning message to include the name of the cmdlet - this will help users running this cmdlet in a script easily find which cmdlet the warning message is coming from. Please do this for all warnings in cmdlets.

File: src/ResourceManager/Insights/Commands.Insights/Alerts/AddAzureRmAlertRuleCommandBase.cs [hunk 3]
```diff
@@ -53,7 +53,7 @@ public abstract class AddAzureRmAlertRuleCommandBase : ManagementCmdletBase
         /// </summary>
         [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The resource group name")]
         [ValidateNotNullOrEmpty]
-        public string ResourceGroup { get; set; }
+        public string ResourceGroupName { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Alerts/AddAzureRmAlertRuleCommandBase.cs]: @gucalder please add an alias to `ResourceGroup` to avoid undocumented breaking changes

File: src/ResourceManager/Insights/Commands.Insights/Alerts/GetAzureRmAlertRuleCommand.cs [hunk 4]
```diff
@@ -41,7 +41,7 @@ public class GetAzureRmAlertRuleCommand : ManagementCmdletBase
         [Parameter(ParameterSetName = GetAzureRmAlertRuleWithNameParamGroup, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The resource group name")]
         [Parameter(ParameterSetName = GetAzureRmAlertRuleWithUriParamGroup, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The resource group name")]
         [ValidateNotNullOrEmpty]
-        public string ResourceGroup { get; set; }
+        public string ResourceGroupName { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Alerts/GetAzureRmAlertRuleCommand.cs]: @gucalder same comment about adding an alias to `ResourceGroup`

File: src/ResourceManager/Insights/Commands.Insights/Autoscale/AddAzureRmAutoscaleSettingCommand.cs [hunk 5]
```diff
@@ -63,7 +64,7 @@ public class AddAzureRmAutoscaleSettingCommand : ManagementCmdletBase
         [Parameter(ParameterSetName = AddAzureRmAutoscaleSettingCreateParamGroup, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The resource group name")]
         [Parameter(ParameterSetName = AddAzureRmAutoscaleSettingUpdateParamGroup, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The resource group name")]
         [ValidateNotNullOrEmpty]
-        public string ResourceGroup { get; set; }
+        public string ResourceGroupName { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Autoscale/AddAzureRmAutoscaleSettingCommand.cs]: @gucalder same comment about adding an alias to `ResourceGroup`

File: src/ResourceManager/Insights/Commands.Insights/Autoscale/GetAzureRmAutoscaleSettingCommand.cs [hunk 6]
```diff
@@ -37,7 +37,7 @@ public class GetAzureRmAutoscaleSettingCommand : ManagementCmdletBase
         /// </summary>
         [Parameter(ParameterSetName = GetAzureRmAutoscaleSettingParamGroup, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The resource group name")]
         [ValidateNotNullOrEmpty]
-        public string ResourceGroup { get; set; }
+        public string ResourceGroupName { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Autoscale/GetAzureRmAutoscaleSettingCommand.cs]: @gucalder same comment about adding an alias to `ResourceGroup`

File: src/ResourceManager/Insights/Commands.Insights/Autoscale/AddAzureRmAutoscaleSettingCommand.cs [hunk 7]
```diff
@@ -41,6 +41,7 @@ public class AddAzureRmAutoscaleSettingCommand : ManagementCmdletBase
         /// </summary>
         [Parameter(ParameterSetName = AddAzureRmAutoscaleSettingUpdateParamGroup, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The complete spec of an AutoscaleSetting")]
         [ValidateNotNullOrEmpty]
+        [Alias("InputObject")]
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Autoscale/AddAzureRmAutoscaleSettingCommand.cs]: @gucalder I would make `InputObject` the actual name of the parameter and add an alias to `SettingSpec`

File: src/ResourceManager/Insights/Commands.Insights/Autoscale/RemoveAzureRmAutoscaleSettingCommand.cs [hunk 8]
```diff
@@ -32,7 +32,7 @@ public class RemoveAzureRmAutoscaleSettingCommand : ManagementCmdletBase
         /// </summary>
         [Parameter(ParameterSetName = RemoveAzureRmAutoscaleSettingParamGroup, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The resource group name")]
         [ValidateNotNullOrEmpty]
-        public string ResourceGroup { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Autoscale/RemoveAzureRmAutoscaleSettingCommand.cs]: @gucalder same comment about adding an alias to `ResourceGroup`

File: src/ResourceManager/Insights/Commands.Insights/LogProfiles/AddAzureRmLogProfileCommand.cs [hunk 9]
```diff
@@ -56,48 +56,57 @@ public class AddAzureRmLogProfileCommand : ManagementCmdletBase
         /// </summary>
         [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The retention in days")]
         [ValidateNotNullOrEmpty]
-        public int? RetentionInDays { get; set; }
+        [Alias("RetentionInDays")]
+        public int? RetentionInDay { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/LogProfiles/AddAzureRmLogProfileCommand.cs]: @gucalder for parameters with units (_e.g.,_ `InDays`, `InMinutes`, etc.), you don't have to worry about making them singluar

File: src/ResourceManager/Insights/Commands.Insights.Test/Utilities.cs (hunk 1 above)
Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights.Test/Utilities.cs]: Removed

File: src/ResourceManager/Insights/Commands.Insights/ActivityLogs/GetAzureRmLogCommand.cs (hunk 2 above)
Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/ActivityLogs/GetAzureRmLogCommand.cs]: Done. I am including the cmdlet name in all warning messages.

File: src/ResourceManager/Insights/Commands.Insights/Alerts/AddAzureRmAlertRuleCommandBase.cs (hunk 3 above)
Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/Alerts/AddAzureRmAlertRuleCommandBase.cs]: I reverted the argument names to ResourceGroup and added an alias ResourceGroupName and removed the exceptions from BreakingChangesIssues.csv

File: src/ResourceManager/Insights/Commands.Insights/Alerts/GetAzureRmAlertRuleCommand.cs (hunk 4 above)
Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/Alerts/GetAzureRmAlertRuleCommand.cs]: Done

File: src/ResourceManager/Insights/Commands.Insights/Autoscale/AddAzureRmAutoscaleSettingCommand.cs (hunk 7 above)
Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/Autoscale/AddAzureRmAutoscaleSettingCommand.cs]: Done.

File: src/ResourceManager/Insights/Commands.Insights/Alerts/AddAzureRmAlertRuleCommandBase.cs (hunk 3 above)
Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/Alerts/AddAzureRmAlertRuleCommandBase.cs]: Alias added... everywhere including documentation

File: src/ResourceManager/Insights/Commands.Insights/Autoscale/AddAzureRmAutoscaleSettingCommand.cs (hunk 5 above)
Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/Autoscale/AddAzureRmAutoscaleSettingCommand.cs]: Done

File: src/ResourceManager/Insights/Commands.Insights/Autoscale/GetAzureRmAutoscaleSettingCommand.cs (hunk 6 above)
Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/Autoscale/GetAzureRmAutoscaleSettingCommand.cs]: Done

File: src/ResourceManager/Insights/Commands.Insights/Autoscale/RemoveAzureRmAutoscaleSettingCommand.cs (hunk 8 above)
Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/Autoscale/RemoveAzureRmAutoscaleSettingCommand.cs]: Done

File: src/ResourceManager/Insights/Commands.Insights/MonitorCmdletBase.cs [hunk 10]
```diff
@@ -33,6 +33,29 @@ public abstract class MonitorCmdletBase : AzureRMCmdlet
...
 
         /// <summary>
+        /// Writes a warning message with the name of the cmdlet, a topic and the message itself
+        /// </summary>
+        /// <param name="topic">The topic, i.e. short description/category, of the message</param>
+        /// <param name="message">The message itself</param>
+        /// <param name="withTimeStamp">true if the message should include a timestamp, false (default) it no timestamp should be included</param>
+        protected void WriteIdentifiedWarning(string topic, string message, bool withTimeStamp = false)
+        {
+            string formattedMessage = Utilities.FormatIdentifiedMessage(
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/MonitorCmdletBase.cs]: @gucalder we are going to be creating our own mechanism for custom obsolete messages and attributes soon, similar to what you are doing here, so we recommend just making this simple for the time being since it will change in the near future ☺️

File: src/ResourceManager/Insights/Commands.Insights/Metrics/GetAzureRmMetricDefinitionCommand.cs [hunk 11]
```diff
@@ -72,6 +73,8 @@ protected string ProcessParameters()
         /// </summary>
         protected override void ProcessRecordInternal()
         {
+            this.WriteIdentifiedWarning("Parameter deprecation", "The DetailedOutput parameter will be deprecated in May 2018.");
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/Metrics/GetAzureRmMetricDefinitionCommand.cs]: @gucalder I would change "May 2018" to "future breaking change release" - even though the next breaking change release _should_ be May 2018, we cannot be 100% positive that is when it is right now

File: src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs [hunk 12]
```diff
@@ -36,11 +36,11 @@ public abstract class LogsCmdletBase : MonitorClientCmdletBase
         private const int MaxNumberOfReturnedRecords = 1000;
         private int MaxRecords = 0;
 
-        internal const string SubscriptionLevelName = "GetBySubscription";
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: @gucalder please revert the change made to these parameter set names - we want parameter sets to be singular strings with pascal casing. I'm not sure if this was just changed because of a merge or not.

Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: Sorry I did not see this before I pushed the changes for the other comments.
Do you need me to make the name of the constant back to what it was?  (I will need to change at least one of them since ResourceGroupName is the name of a parameter now.

Comment (cormacpayne) [reply src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: @gucalder sorry, I made this comment a few hours after when I was looking at a couple of last minute things on this PR. We've received feedback that parameter set names should be a string with no spaces instead of the space-delimited string you are currently using. I recently merged a PR where we fix this in all cmdlets who were using them. If it wouldn't be too much trouble, would you mind revert the set names back to what they were originally, but updating them to reflect the changes you made? That would be great ☺️

Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: Ok. 
The content of each of those constants is now a single string, Pascal casing. I also updated the documentation to reflect that.

Comment (cormacpayne) [reply src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: @gucalder great, let me know when you push that update out

Comment (gucalder) [reply src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: I checked the files and I have already pushed all those changes for parameter groups names: they are all single strings, Pascal casing.

**Comment by cormacpayne** [comment]
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/129/
